staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go

```diff
@@ -0,0 +1,262 @@
+		backend:      backend,
+		buffer:       make(chan *auditinternal.Event, defaultBatchBufferSize),
+		maxBatchSize: defaultBatchMaxSize,
+		maxBatchWait: defaultBatchMaxWait,
+		shutdownCh:   make(chan struct{}),
+		wg:           sync.WaitGroup{},
+		throttle:     flowcontrol.NewTokenBucketRateLimiter(defaultBatchThrottleQPS, defaultBatchThrottleBurst),
+	}
+}
+
+type bufferedBackend struct {
+	backend audit.Backend
+	// Channel to buffer events before underlying backend processing.
+	buffer chan *auditinternal.Event
+	// Maximum number of events that can be consumed at once.
+	maxBatchSize int
+	// Amount of time to wait after consuming events before force consuming another set.
+	//
+	// Receiving maxBatchSize events will always trigger a consume, regardless of
+	// if this amount of time has been reached.
+	maxBatchWait time.Duration
+
+	// Channel to signal that the worker routine to stop and the buffer channel to close.
+	stopCh <-chan struct{}
+
+	// Channel to signal that the worker routine has stopped and therefore
+	// it's safe to assume that no new requests will be initiated.
+	shutdownCh chan struct{}
+
+	// The consuming routine calls WaitGroup.Add before initiating a new
+	// goroutine to consume a request. This goroutine then calls WaitGroup.Done
+	// when completed. The Shutdown method calls WaitGroup.Wait to wait for all
+	// consuming routines exit.
+	wg sync.WaitGroup
+
+	// Limits the number of requests sent to the backend per second.
+	throttle flowcontrol.RateLimiter
+}
+
+var _ audit.Backend = &bufferedBackend{}
+
+func (b *bufferedBackend) Run(stopCh <-chan struct{}) error {
+	b.stopCh = stopCh
+	go func() {
+		// Signal that the working routine has exited.
+		defer close(b.shutdownCh)
+
+		b.runWorkerRoutine()
+
+		// Handle the events that were received after the last buffer
+		// scraping and before this line. Since the buffer is closed, no new
+		// events will come through.
+		for {
+			if last := func() bool {
+				// Recover from any panic in order to try to process all remaining events.
+				// Note, that in case of a panic, the return value will be false and
+				// the loop execution will continue.
+				defer runtime.HandleCrash()
+
+				events := b.collectLastEvents()
+				b.processEvents(events)
+				return len(events) == 0
+			}(); last {
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+func (b *bufferedBackend) Shutdown() {
+	<-b.shutdownCh
+
+	// Wait blocks here until all consuming routines exit.
+	b.wg.Wait()
+}
+
+// runWorkerRoutine runs a loop that collects events from the buffer. When
+// b.stopCh is closed, runWorkerRoutine stops and closes the buffer.
+func (b *bufferedBackend) runWorkerRoutine() {
+	defer close(b.buffer)
+	t := time.NewTimer(b.maxBatchWait)
+	defer t.Stop()
+
+	for {
+		func() {
+			// Recover from any panics caused by this function so a panic in the
+			// goroutine can't bring down the main routine.
+			defer runtime.HandleCrash()
+			defer t.Reset(b.maxBatchWait)
```

Comment:
nit: don't defer this. In practice I don't think it matters, but it's more clear to do it before calling collect events.

Reply:
ok to me